kernel/src/table_changes/log_replay.rs

```diff
@@ -0,0 +1,847 @@
+    add_paths: &'a mut HashSet<String>,
+}
+impl<'a> PreparePhaseVisitor<'a> {
+    fn new(
+        scanner: &'a mut LogReplayScanner,
+        selection_vector: Vec<bool>,
+        add_paths: &'a mut HashSet<String>,
+    ) -> Self {
+        PreparePhaseVisitor {
+            scanner,
+            selection_vector,
+            protocol: None,
+            metadata_info: None,
+            add_paths,
+        }
+    }
+    fn schema() -> DeltaResult<Arc<StructType>> {
+        get_log_schema().project(&[
+            ADD_NAME,
+            REMOVE_NAME,
+            CDC_NAME,
+            COMMIT_INFO_NAME,
+            METADATA_NAME,
+            PROTOCOL_NAME,
+        ])
+    }
+}
+
+impl<'a> RowVisitor for PreparePhaseVisitor<'a> {
+    fn selected_column_names_and_types(
+        &self,
+    ) -> (
+        &'static [crate::expressions::ColumnName],
+        &'static [crate::schema::DataType],
+    ) {
+        // NOTE: The order of the names and types is based on [`PreparePhaseVisitor::schema`]
+        static NAMES_AND_TYPES: LazyLock<ColumnNamesAndTypes> = LazyLock::new(|| {
+            const STRING: DataType = DataType::STRING;
+            const INTEGER: DataType = DataType::INTEGER;
+            const LONG: DataType = DataType::LONG;
+            let string_list: DataType = ArrayType::new(STRING, false).into();
+            let string_string_map = MapType::new(STRING, STRING, false).into();
+            let types_and_names = vec![
+                (STRING, column_name!("add.path")),
+                (STRING, column_name!("remove.path")),
+                (STRING, column_name!("remove.deletionVector.storageType")),
+                (STRING, column_name!("remove.deletionVector.pathOrInlineDv")),
+                (INTEGER, column_name!("remove.deletionVector.offset")),
+                (INTEGER, column_name!("remove.deletionVector.sizeInBytes")),
+                (LONG, column_name!("remove.deletionVector.cardinality")),
+                (STRING, column_name!("cdc.path")),
+                (LONG, column_name!("commitInfo.timestamp")),
+                (STRING, column_name!("metaData.schemaString")),
+                (string_string_map, column_name!("metaData.configuration")),
+                (INTEGER, column_name!("protocol.minReaderVersion")),
+                (INTEGER, column_name!("protocol.minWriterVersion")),
+                (string_list.clone(), column_name!("protocol.readerFeatures")),
+                (string_list, column_name!("protocol.writerFeatures")),
+            ];
+            let (types, names) = types_and_names.into_iter().unzip();
+            (names, types).into()
+        });
+        NAMES_AND_TYPES.as_ref()
+    }
+
+    fn visit<'b>(
+        &mut self,
+        row_count: usize,
+        getters: &[&'b dyn crate::engine_data::GetData<'b>],
+    ) -> DeltaResult<()> {
+        let expected_getters = 15;
+        require!(
+            getters.len() == expected_getters,
```

Comment:
Any particular reason for a named constant here? It's not used anywhere else?
```suggestion
        require!(
            getters.len() == 15,
```

Reply:
True. This was based off `AddRemoveDedupVisitor`, but since we only ever have one case, we don't need it